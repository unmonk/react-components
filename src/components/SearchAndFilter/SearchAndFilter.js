import React, { useState, useEffect, useRef } from "react";
import PropTypes from "prop-types";

import SearchBox from "../SearchBox";
import ContextualMenu from "../ContextualMenu";
import FilterPanelSection from "./FilterPanelSection";

import "./SearchAndFilter.scss";

const SearchAndFilter = ({
  externallyControlled = false,
  onChange,
  filterPanelData,
}) => {
  const [searchTerm, setSearchTerm] = useState("");
  const [filterPanelHidden, setFilterPanelHidden] = useState(true);
  const filterPanelRef = useRef();

  const searchOnChange = (searchTerm) => {
    setSearchTerm(searchTerm);
    if (externallyControlled && onChange) {
      onChange(searchTerm);
    }
  };

  // This useEffect sets up listeners so the panel will close if user clicks
  // anywhere else on the page or hits the escape key
  useEffect(() => {
    const closePanel = () => {
      setFilterPanelHidden(true);
    };

    const mouseDown = (e) => {
      // Check if click is outside of filter panel
      if (!filterPanelRef?.current?.contains(e.target)) {
        // If so, close the panel
        closePanel();
      }
    };

    const keyDown = (e) => {
      if (e.code === "Escape") {
        // Close panel if Esc keydown detected
        closePanel();
      }
    };

    // Add listener on document to capture click events
    document.addEventListener("mousedown", mouseDown);
    // Add listener on document to capture keydown events
    document.addEventListener("keydown", keyDown);
    // return function to be called when unmounted
    return () => {
      document.removeEventListener("mousedown", mouseDown);
      document.removeEventListener("keydown", keyDown);
    };
  }, []);

  return (
    <div className="search-and-filter">
      <SearchBox
        autocomplete="off"
        externallyControlled={externallyControlled}
        placeholder="Search and filter"
        onChange={(searchTerm) => searchOnChange(searchTerm)}
        onClick={() => setFilterPanelHidden(false)}
        onFocus={() => setFilterPanelHidden(false)}
        value={searchTerm}
      />
      {filterPanelData && (
        <div
          className="search-and-filter__panel"
          ref={filterPanelRef}
          aria-hidden={filterPanelHidden}
        >
          <ContextualMenu>
            {filterPanelData.map((filterPanelSectionData) => {
              return (
                <div key={filterPanelSectionData.id}>
                  <FilterPanelSection data={filterPanelSectionData} />
                </div>
              );
            })}
          </ContextualMenu>
        </div>
      )}
    </div>
  );
};

SearchAndFilter.propTypes = {
  externallyControlled: PropTypes.bool,
  filterPanelData: PropTypes.arrayOf(
    PropTypes.shape({
      heading: PropTypes.string,
      chips: PropTypes.arrayOf(
        PropTypes.shape({
          id: PropTypes.number,
          lead: PropTypes.string,
          value: PropTypes.string,
        })
      ),
    })
  ),
  onChange: PropTypes.func,
  onSubmit: PropTypes.func,
};

export default SearchAndFilter;
