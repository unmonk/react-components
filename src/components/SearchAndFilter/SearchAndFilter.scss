.search-and-filter {
  span[class*="p-contextual-menu"],
  .p-contextual-menu__dropdown {
    display: block;
    max-width: 100%;
    width: 100%;
  }

  .p-contextual-menu__dropdown {
    padding: 0.5rem;
  }

  .p-search-box {
    margin: 0;
  }

  .p-contextual-menu__dropdown {
    padding: 1rem;
    top: 0;
    left: 0;
  }

  &__panel {
    padding: 0;
    transition: opacity 0.25s;

    &[aria-hidden="true"] {
      opacity: 0;
      pointer-events: none;
    }
  }
}
