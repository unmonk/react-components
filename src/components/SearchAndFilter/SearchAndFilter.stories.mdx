import { Meta, Story, Preview, Props } from "@storybook/addon-docs/blocks";
import { withKnobs } from "@storybook/addon-knobs";
import SearchAndFilter from "./SearchAndFilter";

<Meta
  title="SearchAndFilter"
  component={SearchAndFilter}
  decorators={[withKnobs]}
/>

### Search And Filter

This component is used to search and filter large datasets.

### Props

<Props of={SearchAndFilter} />

### Default

<Preview>
  <Story name="Default">
    <SearchAndFilter />
  </Story>
</Preview>

### External state

If you wish to control the value of the data set via external state you can set
the `externallyControlled` prop and provide an `onChange` method to update the
state and the `value` from state.

<Preview>
  <Story name="External state">
    <SearchAndFilter
      externallyControlled
      onChange={(searchTerm) => {
        console.log("onChange: ", searchTerm);
      }}
      filterPanelData={[
        {
          id: 0,
          heading: "Clouds",
          chips: [{ value: "Google" }, { value: "AWS" }, { value: "Azure" }],
        },
        {
          id: 1,
          heading: "Regions",
          chips: [
            { value: "us-east1" },
            { value: "us-east2" },
            { value: "us-east3" },
          ],
        },
        {
          id: 2,
          heading: "Owner",
          chips: [{ value: "foo" }, { value: "bar" }, { value: "baz" }],
        },
      ]}
    />
  </Story>
</Preview>
