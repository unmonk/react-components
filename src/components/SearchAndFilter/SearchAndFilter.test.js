import { shallow, mount } from "enzyme";
import React from "react";

import SearchAndFilter from "./SearchAndFilter";

const sampleData = [
  {
    id: 0,
    heading: "Clouds",
    chips: [{ value: "Google" }, { value: "AWS" }, { value: "Azure" }],
  },
  {
    id: 1,
    heading: "Regions",
    chips: [
      { value: "us-east1" },
      { value: "us-east2" },
      { value: "us-east3" },
    ],
  },
  {
    id: 3,
    heading: "Owner",
    chips: [{ value: "foo" }, { value: "bar" }, { value: "baz" }],
  },
];

describe("Search and filter", () => {
  it("renders", () => {
    const wrapper = shallow(<SearchAndFilter />);
    expect(wrapper).toMatchSnapshot();
  });

  it("hide the clear button when there is no value in search box", () => {
    const wrapper = mount(<SearchAndFilter />);
    expect(wrapper.find(".p-search-box__reset").exists()).toBe(false);
  });

  it("onChange handler called", () => {
    const mockOnChange = jest.fn();
    const wrapper = mount(
      <SearchAndFilter externallyControlled onChange={mockOnChange} />
    );
    wrapper.find(".p-search-box__input").simulate("change");
    expect(mockOnChange).toHaveBeenCalledTimes(1);
  });

  it("shows panel on focus", () => {
    const mockOnChange = jest.fn();
    const wrapper = mount(
      <SearchAndFilter
        externallyControlled
        onChange={mockOnChange}
        filterPanelData={sampleData}
      />
    );
    expect(
      wrapper.find(".search-and-filter__panel").prop("aria-hidden")
    ).toEqual(true);
    wrapper.find(".p-search-box__input").simulate("focus");
    expect(
      wrapper.find(".search-and-filter__panel").prop("aria-hidden")
    ).toEqual(false);
  });

  it("shows panel on click", () => {
    const mockOnChange = jest.fn();
    const wrapper = mount(
      <SearchAndFilter
        externallyControlled
        onChange={mockOnChange}
        filterPanelData={sampleData}
      />
    );
    expect(
      wrapper.find(".search-and-filter__panel").prop("aria-hidden")
    ).toEqual(true);
    wrapper.find(".p-search-box__input").simulate("click");
    expect(
      wrapper.find(".search-and-filter__panel").prop("aria-hidden")
    ).toEqual(false);
  });
});
