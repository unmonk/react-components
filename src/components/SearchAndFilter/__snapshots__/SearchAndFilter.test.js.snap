// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search and filter renders 1`] = `
<div
  className="search-and-filter"
>
  <SearchBox
    autocomplete="off"
    externallyControlled={false}
    onChange={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    placeholder="Search and filter"
    value=""
  />
</div>
`;
